central/resourcecollection/datastore/datastore_impl.go

```diff
@@ -427,3 +431,156 @@ func verifyCollectionObjectNotEmpty(obj *storage.ResourceCollection) error {
 	}
 	return nil
 }
+
+func (ds *datastoreImpl) ResolveListDeployments(ctx context.Context, collection *storage.ResourceCollection) ([]*storage.ListDeployment, error) {
+
+	if err := verifyCollectionConstraints(collection); err != nil {
+		return nil, err
+	}
+
+	query, err := ds.resolveCollectionQuery(ctx, collection)
+	if err != nil {
+		return nil, err
+	}
+	return ds.deploymentDS.SearchListDeployments(ctx, query)
+}
+
+func (ds *datastoreImpl) resolveCollectionQuery(ctx context.Context, collection *storage.ResourceCollection) (*v1.Query, error) {
+	var collections []*storage.ResourceCollection
+	var collectionSet set.Set[string]
+	var disjunctions []*v1.Query
+
+	collections = append(collections, collection)
+
+	ds.lock.RLock()
+	defer ds.lock.RUnlock()
+
+	for len(collections) > 0 {
+
+		// get first index and remove from list
+		collection := collections[0]
+		collections = collections[1:]
+
+		if !collectionSet.Add(collection.GetId()) {
+			continue
+		}
+
+		// resolve the collection to queries
+		queries, err := collectionToQueries(collection)
+		if err != nil {
+			return nil, err
+		}
+		disjunctions = append(disjunctions, queries...)
+
+		// add embedded values
+		embeddedList, _, err := ds.storage.GetMany(ctx, embeddedCollectionsToIDList(collection.GetEmbeddedCollections()))
+		if err != nil {
+			return nil, err
+		}
+		collections = append(collections, embeddedList...)
+	}
+
+	return pkgSearch.DisjunctionQuery(disjunctions...), nil
+}
+
+// collectionToQueries returns a list of queries derived from the given resource collection's storage.ResourceSelector list
+// these should be combined as disjunct with any resolved embedded queries
+func collectionToQueries(collection *storage.ResourceCollection) ([]*v1.Query, error) {
+	var ret []*v1.Query
+
+	for _, resourceSelector := range collection.GetResourceSelectors() {
+		var selectorRuleQueries []*v1.Query
+		for _, selectorRule := range resourceSelector.GetRules() {
+
+			// currently we only support disjunction operations here
+			if selectorRule.GetOperator() != storage.BooleanOperator_OR {
+				return nil, errors.Wrapf(errox.InvalidArgs, "%q boolean operator unsupported", selectorRule.GetOperator().String())
+			}
+
+			fieldLabel, present := supportedFieldNames[selectorRule.GetFieldName()]
+			if !present {
+				return nil, errors.Wrapf(errox.InvalidArgs, "unsupported field label %q", selectorRule.GetFieldName())
+			}
```

Comment:
I think it would be best to do this refactor in a future PR and to instead get the functionality in.
In the future I envision the flow for these functions to exclude a standalone collection resolver, and have the flow logically merged into the functions, i.e. add/dryrun/update

Reply:
`verifyCollectionConstraints` only needs to be called on the passed in collection because we can assume that every collection object in the datastore is a valid one. The reason all of these checks are here is because it's better to check every object for posterity than it would be to ignore all of the potential edge cases you might encounter when expanding embedded objects retrieved from the datastore.

`collectionToQueries` is called on every collection object, including the embedded ones, so the functionality of it should stay generic in nature and just focus on manipulating already sanitized objects.